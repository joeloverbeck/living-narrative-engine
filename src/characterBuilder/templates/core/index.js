/**
 * @file Core template exports
 * @module characterBuilder/templates/core
 */

// Page template container
export { createCharacterBuilderPage } from './pageTemplate.js';

// Header template and utilities (HTMLTEMP-003)
export {
  createHeader,
  createBreadcrumb,
  createHeaderSearch,
} from './headerTemplate.js';

// NOTE: These exports reference files that will be created in subsequent tickets
export { createMain } from './mainTemplate.js';
export { createFooter } from './footerTemplate.js';
export { createModal } from './modalTemplate.js';
