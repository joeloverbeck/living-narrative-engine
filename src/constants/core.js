export const CORE_MOD_ID = 'core';

// Allow tests or specialized builds to override the default cap via an
// environment variable. This keeps heavy loops short for test suites.
const parsedMax = parseInt(process.env.MAX_AVAILABLE_ACTIONS_PER_TURN, 10);
export const MAX_AVAILABLE_ACTIONS_PER_TURN =
  Number.isInteger(parsedMax) && parsedMax > 0 ? parsedMax : 30000;
