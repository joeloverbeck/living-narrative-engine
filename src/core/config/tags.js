// src/core/tags.js

/**
 * Array containing the tag used to identify systems that require initialization.
 * @constant
 * @type {string[]}
 */
export const INITIALIZABLE = ['initializableSystem'];
export const SHUTDOWNABLE = ['shutdownable']