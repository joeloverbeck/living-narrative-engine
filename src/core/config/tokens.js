// src/core/config/tokens.js

/**
 * @fileoverview Centralized repository for Dependency Injection (DI) keys/tokens.
 * Using tokens instead of raw strings prevents typos and aids refactoring.
 */

/**
 * A frozen object containing all unique keys used for registering and resolving
 * services and dependencies within the application's DI container.
 * Using Object.freeze prevents accidental modification.
 *
 * @typedef {string} DiToken
 *
 * @property {DiToken} ILogger - Token for the core logging service.
 * @property {DiToken} EventBus - Token for the central event bus (legacy).
 * @property {DiToken} IDataFetcher - Token for the data fetching service.
 * @property {DiToken} IConfiguration - Token for static configuration access.
 * @property {DiToken} IPathResolver - Token for resolving data paths.
 * @property {DiToken} ISchemaValidator - Token for validating data against schemas.
 * @property {DiToken} IDataRegistry - Token for storing loaded game data.
 * @property {DiToken} ISpatialIndexManager - Token for managing the spatial index.
 *
 * --- External Dependencies / Environment ---
 * @property {DiToken} WindowDocument - Token for the browser's global `document` object.
 * @property {DiToken} outputDiv - Token for the main output HTML element (legacy/direct access).
 * @property {DiToken} inputElement - Token for the command input HTML element (legacy/direct access).
 * @property {DiToken} titleElement - Token for the title HTML element (legacy/direct access).
 *
 * --- DOM UI Layer (Refactored) ---
 * @property {DiToken} IDocumentContext - Token for the DOM access abstraction service.
 * @property {DiToken} DomElementFactory - Token for the utility creating DOM elements.
 * @property {DiToken} UiMessageRenderer - Token for the component rendering UI messages/echoes.
 * @property {DiToken} TitleRenderer - Token for the component rendering the main H1 title.
 * @property {DiToken} InputStateController - Token for the component controlling input field state.
 * @property {DiToken} LocationRenderer - Token for the component rendering location details.
 * @property {DiToken} InventoryPanel - Token for the component managing the inventory panel UI.
 * @property {DiToken} ActionButtonsRenderer - Token for the component rendering available action buttons.
 * @property {DiToken} PerceptionLogRenderer - Token for the component rendering perception logs. // <<< ADDED
 * @property {DiToken} DomUiFacade - Token for the facade aggregating all UI components.
 * @property {DiToken} DomRenderer - Token for the legacy DOM rendering class (deprecated).
 *
 * --- Loaders ---
 * @property {DiToken} SchemaLoader - Token for the schema loading service.
 * @property {DiToken} ManifestLoader - Token for the manifest loading service (deprecated?).
 * @property {DiToken} RuleLoader - Token for the rule loading service.
 * @property {DiToken} ComponentDefinitionLoader - Token for loading component definitions.
 * @property {DiToken} ActionLoader - Token for the action loading service.
 * @property {DiToken} EventLoader - Token for the event loading service.
 * @property {DiToken} EntityLoader - Token for loading entity definitions.
 * @property {DiToken} WorldLoader - Token for orchestrating world data loading.
 * @property {DiToken} GameConfigLoader - Token for loading the main game configuration file.
 * @property {DiToken} ModManifestLoader - Token for loading mod manifests.
 *
 * --- Core Services & Managers (Implementations - some will be replaced by Interface Tokens below) ---
 * @property {DiToken} GameDataRepository - Token for accessing registered game data (implementation).
 * @property {DiToken} EntityManager - Token for managing game entities and components (implementation).
 * @property {DiToken} TargetResolutionService - Token for resolving action targets.
 * @property {DiToken} JsonLogicEvaluationService - Token for evaluating JsonLogic rules.
 * @property {DiToken} ActionValidationContextBuilder - Token for building action validation contexts.
 * @property {DiToken} PrerequisiteEvaluationService - Token for evaluating action/quest prerequisites.
 * @property {DiToken} DomainContextCompatibilityChecker - Token for checking domain context compatibility.
 * @property {DiToken} ActionValidationService - Token for validating actions.
 * @property {DiToken} PayloadValueResolverService - Token for resolving payload values.
 * @property {DiToken} TurnHandlerResolver - Token for the service that resolves the correct turn handler.
 * @property {DiToken} PlayerTurnHandler - Token for the player-specific turn handler implementation.
 * @property {DiToken} AITurnHandler - Token for the AI-specific turn handler implementation.
 * @property {DiToken} SystemServiceRegistry - Token for the registry mapping system IDs to services.
 * @property {DiToken} SystemDataRegistry - Token for the registry mapping system IDs to data sources.
 * @property {DiToken} PlayerPromptService - Token for the service managing player action prompting (implementation).
 * @property {DiToken} CommandOutcomeInterpreter - Token for the service interpreting command outcomes (implementation).
 * @property {DiToken} SubscriptionLifecycleManager - Token for managing subscription lifecycles (implementation).
 * @property {DiToken} PerceptionUpdateService - Token for the service updating perception logs.
 *
 * --- Core Service Interfaces ---
 * @property {DiToken} ISafeEventDispatcher - Token for the safe event dispatching utility interface.
 * @property {DiToken} IValidatedEventDispatcher - Token for dispatching validated events interface.
 * @property {DiToken} IActionExecutor - Token for executing game actions interface.
 * @property {DiToken} IWorldContext - Token for managing the overall world context interface.
 * @property {DiToken} ICommandParser - Token for parsing player commands interface.
 * @property {DiToken} ICommandProcessor - Token for the command processing service interface.
 * @property {DiToken} IActionDiscoverySystem - Token for the action discovery system interface.
 * @property {DiToken} IInputHandler - Token for handling player input interface.
 * @property {DiToken} ITurnOrderService - Token for the turn order management service interface.
 * @property {DiToken} ITurnManager - Token for the turn management service interface.
 * @property {DiToken} ICommandInputPort - Token for the command input port interface.
 * @property {DiToken} IPromptOutputPort - Token for the prompt output port interface.
 * @property {DiToken} ITurnEndPort - Token for the turn end port interface.
 * @property {DiToken} IPlayerPromptService - Token for the player prompt service interface.
 * @property {DiToken} ICommandOutcomeInterpreter - Token for the command outcome interpreter interface.
 * @property {DiToken} IEntityManager - Token for the entity manager interface.
 * @property {DiToken} IGameDataRepository - Token for the game data repository interface.
 *
 * --- Initialization & Orchestration ---
 * @property {DiToken} WorldInitializer - Token for initializing the game world.
 * @property {DiToken} SystemInitializer - Token for initializing tagged systems.
 * @property {DiToken} InitializationService - Token for the main initialization orchestration service.
 * @property {DiToken} ShutdownService - Token for the main shutdown orchestration service.
 * @property {DiToken} GameLoop - Token for the main game loop.
 * @property {DiToken} InputSetupService - Token for setting up input handling.
 *
 *
 * --- Logic/Interpretation Layer ---
 * @property {DiToken} OperationRegistry - Token for the operation handler registry service.
 * @property {DiToken} OperationInterpreter - Token for the operation interpreter service.
 * @property {DiToken} SystemLogicInterpreter - Token for the system logic interpreter service.
 *
 * --- Operation Handlers (Registered within Interpreter bundle) ---
 * @property {DiToken} DispatchEventHandler - Token for the 'DISPATCH_EVENT' operation handler.
 * @property {DiToken} LogHandler - Token for the 'LOG' operation handler.
 * @property {DiToken} ModifyComponentHandler - Token for the 'MODIFY_COMPONENT' operation handler.
 * @property {DiToken} AddComponentHandler - Token for the 'ADD_COMPONENT' operation handler.
 * @property {DiToken} RemoveComponentHandler - Token for the 'REMOVE_COMPONENT' operation handler.
 * @property {DiToken} QueryComponentHandler - Token for the 'QUERY_COMPONENT' operation handler.
 * @property {DiToken} SetVariableHandler - Token for the 'SET_VARIABLE' operation handler.
 * @property {DiToken} QuerySystemDataHandler - Token for the 'QUERY_SYSTEM_DATA' operation handler.
 */
export const tokens = Object.freeze({
    // Core Interfaces/Abstractions & Externals
    ILogger: 'ILogger',
    EventBus: 'EventBus', // Legacy
    IDataFetcher: 'IDataFetcher',
    IConfiguration: 'IConfiguration',
    IPathResolver: 'IPathResolver',
    ISchemaValidator: 'ISchemaValidator',
    IDataRegistry: 'IDataRegistry',
    ISpatialIndexManager: 'ISpatialIndexManager',

    // --- External Dependencies / Environment ---
    WindowDocument: 'WindowDocument',
    outputDiv: 'outputDiv', // Legacy/Direct access
    inputElement: 'inputElement', // Legacy/Direct access
    titleElement: 'titleElement', // Legacy/Direct access

    // --- DOM UI Layer (Refactored) ---
    IDocumentContext: 'IDocumentContext',
    DomElementFactory: 'DomElementFactory',
    UiMessageRenderer: 'UiMessageRenderer',
    TitleRenderer: 'TitleRenderer',
    InputStateController: 'InputStateController',
    LocationRenderer: 'LocationRenderer',
    InventoryPanel: 'InventoryPanel',
    ActionButtonsRenderer: 'ActionButtonsRenderer',
    PerceptionLogRenderer: 'PerceptionLogRenderer', // <<< ADDED
    DomUiFacade: 'DomUiFacade',

    // Loaders
    SchemaLoader: 'SchemaLoader',
    RuleLoader: 'RuleLoader',
    ComponentDefinitionLoader: 'ComponentDefinitionLoader',
    ActionLoader: 'ActionLoader',
    EventLoader: 'EventLoader',
    EntityLoader: 'EntityLoader',
    WorldLoader: 'WorldLoader',
    GameConfigLoader: 'GameConfigLoader',
    ModManifestLoader: 'ModManifestLoader',

    // Core Services & Managers (Concrete Implementations - some may be deprecated for interface tokens)
    GameDataRepository: 'GameDataRepository', // Concrete class token
    EntityManager: 'EntityManager',           // Concrete class token
    TargetResolutionService: 'TargetResolutionService',
    JsonLogicEvaluationService: 'JsonLogicEvaluationService',
    ActionValidationContextBuilder: 'ActionValidationContextBuilder',
    PrerequisiteEvaluationService: 'PrerequisiteEvaluationService',
    DomainContextCompatibilityChecker: 'DomainContextCompatibilityChecker',
    ActionValidationService: 'ActionValidationService',
    TurnHandlerResolver: 'TurnHandlerResolver',
    PlayerTurnHandler: 'PlayerTurnHandler',
    AITurnHandler: 'AITurnHandler',
    SystemServiceRegistry: 'SystemServiceRegistry',
    SystemDataRegistry: 'SystemDataRegistry',
    PlayerPromptService: 'PlayerPromptService',         // Concrete class token
    CommandOutcomeInterpreter: 'CommandOutcomeInterpreter', // Concrete class token
    SubscriptionLifecycleManager: 'SubscriptionLifecycleManager', // Concrete class token
    PerceptionUpdateService: 'PerceptionUpdateService', // <<< ADDED TOKEN (already present from previous user request, good)

    // Core Service Interfaces
    ISafeEventDispatcher: 'ISafeEventDispatcher',
    IValidatedEventDispatcher: 'IValidatedEventDispatcher',
    IActionExecutor: 'IActionExecutor',
    IWorldContext: 'IWorldContext',
    ICommandParser: 'ICommandParser',
    ICommandProcessor: 'ICommandProcessor',
    IActionDiscoverySystem: 'IActionDiscoverySystem',
    IInputHandler: 'IInputHandler', // May become obsolete
    ITurnOrderService: 'ITurnOrderService',
    ITurnManager: 'ITurnManager',
    ITurnContext: 'ITurnContext',
    ICommandInputPort: 'ICommandInputPort',
    IPromptOutputPort: 'IPromptOutputPort',
    ITurnEndPort: 'ITurnEndPort',
    IPlayerPromptService: 'IPlayerPromptService',             // Interface token
    ICommandOutcomeInterpreter: 'ICommandOutcomeInterpreter', // Interface token
    IEntityManager: 'IEntityManager',                         // Interface token
    IGameDataRepository: 'IGameDataRepository',             // Interface token


    // Initialization & Orchestration
    WorldInitializer: 'WorldInitializer',
    SystemInitializer: 'SystemInitializer',
    InitializationService: 'InitializationService',
    ShutdownService: 'ShutdownService',
    GameLoop: 'GameLoop',
    InputSetupService: 'InputSetupService',


    // Logic/Interpretation Layer
    OperationRegistry: 'OperationRegistry',
    OperationInterpreter: 'OperationInterpreter',
    SystemLogicInterpreter: 'SystemLogicInterpreter',

    // Operation Handlers (Registered within Interpreter bundle)
    DispatchEventHandler: 'DispatchEventHandler',
    LogHandler: 'LogHandler',
    ModifyComponentHandler: 'ModifyComponentHandler',
    AddComponentHandler: 'AddComponentHandler',
    RemoveComponentHandler: 'RemoveComponentHandler',
    QueryComponentHandler: 'QueryComponentHandler',
    SetVariableHandler: 'SetVariableHandler',
    QuerySystemDataHandler: 'QuerySystemDataHandler',
});
// --- FILE END ---