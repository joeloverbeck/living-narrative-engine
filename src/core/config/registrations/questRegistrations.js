// src/core/config/registrations/questRegistrations.js

/**
 * @fileoverview Registers quest-related services and systems.
 */

// --- JSDoc Imports for Type Hinting ---
/** @typedef {import('../../appContainer.js').default} AppContainer */
/** @typedef {import('../../interfaces/coreServices.js').ILogger} ILogger */

// --- Service/System Imports ---
import {QuestPrerequisiteService} from '../../../services/questPrerequisiteService.js';
import {QuestRewardService} from '../../../services/questRewardService.js';
import {ObjectiveEventListenerService} from '../../../services/objectiveEventListenerService.js';
import {ObjectiveStateCheckerService} from '../../../services/objectiveStateCheckerService.js';
import QuestSystem from '../../../systems/questSystem.js';
import {QuestStartTriggerSystem} from '../../../systems/questStartTriggerSystem.js';

// --- DI & Helper Imports ---
import {tokens} from '../../tokens.js';
import {Registrar} from '../../dependencyInjection/registrarHelpers.js';
import {INITIALIZABLE} from "../../tags";

// --- Constant Tag ---

/**
 * Registers quest-related services and systems.
 *
 * @export
 * @param {AppContainer} container - The application's DI container.
 */
export function registerQuestSystems(container) {
    const registrar = new Registrar(container);
    /** @type {ILogger} */
    const logger = container.resolve(tokens.ILogger);
    logger.debug('Quest Registration: Starting...');

    // === Quest Services (No Initialization Tag) ===
    registrar.single(tokens.QuestPrerequisiteService, QuestPrerequisiteService, []);
    logger.debug(`Quest Registration: Registered ${tokens.QuestPrerequisiteService}.`);

    registrar.single(tokens.QuestRewardService, QuestRewardService, [
        tokens.GameDataRepository,
        tokens.GameStateManager,
        tokens.ValidatedEventDispatcher,
        tokens.ILogger
    ]);
    logger.debug(`Quest Registration: Registered ${tokens.QuestRewardService}.`);

    registrar.single(tokens.ObjectiveEventListenerService, ObjectiveEventListenerService, [
        tokens.EventBus,
        tokens.GameDataRepository
    ]);
    logger.debug(`Quest Registration: Registered ${tokens.ObjectiveEventListenerService}.`);

    registrar.single(tokens.ObjectiveStateCheckerService, ObjectiveStateCheckerService, [
        tokens.EventBus,
        tokens.GameDataRepository,
        tokens.EntityManager,
        tokens.GameStateManager
    ]);
    logger.debug(`Quest Registration: Registered ${tokens.ObjectiveStateCheckerService}.`);

    // === Quest Systems (Tagged as Initializable) ===

    // FIX: Use singletonFactory because QuestSystem constructor expects a single object argument
    registrar.tagged(INITIALIZABLE).singletonFactory(tokens.QuestSystem, (c) => {
        // Resolve all dependencies
        const gameDataRepository = c.resolve(tokens.GameDataRepository);
        const eventBus = c.resolve(tokens.EventBus);
        const entityManager = c.resolve(tokens.EntityManager);
        const gameStateManager = c.resolve(tokens.GameStateManager);
        const questPrerequisiteService = c.resolve(tokens.QuestPrerequisiteService);
        const questRewardService = c.resolve(tokens.QuestRewardService);
        const objectiveEventListenerService = c.resolve(tokens.ObjectiveEventListenerService);
        const objectiveStateCheckerService = c.resolve(tokens.ObjectiveStateCheckerService);

        // Pass dependencies as a single object matching the constructor's expected destructuring
        return new QuestSystem({
            gameDataRepository,
            eventBus,
            entityManager,
            gameStateManager,
            questPrerequisiteService,
            questRewardService,
            objectiveEventListenerService,
            objectiveStateCheckerService
        });
    });
    logger.debug(`Quest Registration: Registered ${tokens.QuestSystem} tagged with ${INITIALIZABLE.join(', ')}.`);

    // QuestStartTriggerSystem likely takes individual args, so .single is probably fine (verify its constructor if needed)
    registrar.tagged(INITIALIZABLE).single(tokens.QuestStartTriggerSystem, QuestStartTriggerSystem, [
        tokens.EventBus,
        tokens.GameStateManager,
        tokens.GameDataRepository
    ]);
    logger.debug(`Quest Registration: Registered ${tokens.QuestStartTriggerSystem} tagged with ${INITIALIZABLE.join(', ')}.`);

    const registrationCount = 6;
    logger.info(`Quest Registration: Completed registering ${registrationCount} quest services/systems.`);
}