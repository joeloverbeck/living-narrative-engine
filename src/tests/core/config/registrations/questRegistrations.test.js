// src/core/config/registrations/questRegistrations.test.js

// --- JSDoc Imports for Type Hinting ---
/** @typedef {import('../../interfaces/coreServices.js').ILogger} ILogger */
/** @typedef {import('../../interfaces/coreServices.js').IEventBus} IEventBus */
/** @typedef {import('../../../entities/entityManager.js').default} EntityManager */
/** @typedef {import('../../gameStateManager.js').default} GameStateManager */
/** @typedef {import('../../../services/questPrerequisiteService.js').QuestPrerequisiteService} QuestPrerequisiteService */
/** @typedef {import('../../../services/questRewardService.js').QuestRewardService} QuestRewardService */
/** @typedef {import('../../../services/objectiveEventListenerService.js').ObjectiveEventListenerService} ObjectiveEventListenerService */
/** @typedef {import('../../../services/objectiveStateCheckerService.js').ObjectiveStateCheckerService} ObjectiveStateCheckerService */
/** @typedef {import('../../services/gameDataRepository.js').GameDataRepository} GameDataRepository */
/** @typedef {import('../../services/validatedEventDispatcher.js').default} ValidatedEventDispatcher */ // Added for mock
/** @typedef {any} AppContainer */ // Using 'any' for simplicity

// --- Jest Imports ---
import {describe, beforeEach, it, expect, jest} from '@jest/globals';

// --- Class Under Test ---
import {registerQuestSystems} from '../../../../core/config/registrations/questRegistrations.js'; // Adjust path

// --- Dependencies ---
import {tokens} from '../../../../core/tokens.js';
import {INITIALIZABLE} from "../../../../core/tags";

// --- Mock Implementations ---
const mockLogger = {info: jest.fn(), warn: jest.fn(), error: jest.fn(), debug: jest.fn()};
const mockEventBus = {on: jest.fn(), off: jest.fn(), emit: jest.fn()};
const mockEntityManager = { /* Methods added if needed by factories */};
const mockGameStateManager = { /* Methods added if needed by factories */};
const mockQuestPrerequisiteService = { /* Methods added if needed */};
const mockQuestRewardService = { /* Methods added if needed */};
const mockObjectiveEventListenerService = { /* Methods added if needed */};
const mockObjectiveStateCheckerService = { /* Methods added if needed */};

// FIX: Provide a more complete mock for GameDataRepository
const mockGameDataRepository = {
    // Add methods expected by QuestSystem or its dependencies' constructors
    // Based on the provided GameDataRepository code:
    getActionDefinition: jest.fn(),
    getEntityDefinition: jest.fn(),
    getAllActionDefinitions: jest.fn(() => []), // Return empty array for safety
    getAllEntityDefinitions: jest.fn(() => []), // Return empty array for safety
    // Add any other methods that might be checked or used during construction
};

const mockvalidatedEventDispatcher = {dispatch: jest.fn()};

// --- Mock DI Container ---
const createMockContainer = () => {
    const registrations = new Map();
    const instances = new Map(); // To handle singleton resolution

    const container = {
        _registrations: registrations,
        register: jest.fn((token, factoryOrValue, options = {}) => {
            if (!token) throw new Error('Mock Register Error: Token is required.');
            registrations.set(token, {factoryOrValue, options});
            // console.log(`Mock Register: ${String(token)} with options ${JSON.stringify(options)}`);
        }),
        resolve: jest.fn((token) => {
            // console.log(`Mock Resolve Request: ${String(token)}`);
            // Always return predefined mocks first
            if (token === tokens.ILogger) return mockLogger;
            if (token === tokens.EventBus) return mockEventBus;
            if (token === tokens.EntityManager) return mockEntityManager;
            if (token === tokens.GameStateManager) return mockGameStateManager;
            if (token === tokens.QuestPrerequisiteService && instances.has(token)) return instances.get(token);
            if (token === tokens.QuestPrerequisiteService) return mockQuestPrerequisiteService;
            if (token === tokens.QuestRewardService) return mockQuestRewardService;
            if (token === tokens.ObjectiveEventListenerService) return mockObjectiveEventListenerService;
            if (token === tokens.ObjectiveStateCheckerService) return mockObjectiveStateCheckerService;
            // Use the improved mock here
            if (token === tokens.GameDataRepository) return mockGameDataRepository;
            if (token === tokens.ValidatedEventDispatcher) return mockvalidatedEventDispatcher;


            const registration = registrations.get(token);
            if (!registration) {
                // Add other known dependencies if needed for complex factory resolutions
                // if (token === tokens.SomeOtherService) return mockSomeOtherService;
                throw new Error(`Mock Resolve Error: Token not registered or mocked: ${String(token)}`);
            }

            // Handle singleton resolution: create instance once if factory exists
            if (registration.options?.lifecycle === 'singleton') {
                if (instances.has(token)) {
                    // console.log(`Mock Resolve: Returning existing singleton instance for ${String(token)}`);
                    return instances.get(token);
                }
                if (typeof registration.factoryOrValue === 'function') {
                    // console.log(`Mock Resolve: Creating singleton instance for ${String(token)}`);
                    // Execute factory, passing the container itself
                    const instance = registration.factoryOrValue(container);
                    instances.set(token, instance);
                    return instance;
                } else {
                    // It's a pre-registered instance
                    instances.set(token, registration.factoryOrValue);
                    return registration.factoryOrValue;
                }
            }

            // Handle transient or other lifecycles (basic execution)
            if (typeof registration.factoryOrValue === 'function') {
                // console.log(`Mock Resolve: Executing factory for ${String(token)}`);
                return registration.factoryOrValue(container);
            }

            // console.log(`Mock Resolve: Returning raw value for ${String(token)}`);
            return registration.factoryOrValue; // Return raw value/class
        }),
        resolveAll: jest.fn((tag) => {
            // console.log(`Mock ResolveAll Request: ${tag}`);
            const found = [];
            for (const [token, reg] of registrations.entries()) {
                if (reg.options?.tags?.includes(tag)) {
                    // Return the token itself for simplicity in this test setup
                    found.push(token);
                }
            }
            // console.log(`Mock ResolveAll Found: ${found.map(String)}`);
            return found;
        })
    };
    return container;
};


describe('registerQuestSystems', () => {
    /** @type {ReturnType<typeof createMockContainer>} */
    let mockContainer;

    const questTokens = [
        tokens.QuestPrerequisiteService,
        tokens.QuestRewardService,
        tokens.ObjectiveEventListenerService,
        tokens.ObjectiveStateCheckerService,
        tokens.QuestSystem,
        tokens.QuestStartTriggerSystem,
    ];
    const initializableQuestTokens = [
        tokens.QuestSystem,
        tokens.QuestStartTriggerSystem,
    ];
    const nonInitializableQuestTokens = questTokens.filter(t => !initializableQuestTokens.includes(t));


    beforeEach(() => {
        jest.clearAllMocks();
        mockContainer = createMockContainer();
        // Pre-register essential dependency needed *by* the registration function itself
        mockContainer.register(tokens.ILogger, mockLogger, {lifecycle: 'singleton'});
    });

    it('should register all 6 quest services/systems', () => {
        registerQuestSystems(mockContainer);

        expect(mockContainer.register).toHaveBeenCalledTimes(questTokens.length + 1); // +1 for ILogger

        questTokens.forEach(token => {
            expect(mockContainer.register).toHaveBeenCalledWith(
                token,
                expect.any(Function), // All registered via factories/single
                expect.objectContaining({lifecycle: 'singleton'})
            );
        });
    });

    it(`should tag ${initializableQuestTokens.length} systems with '${INITIALIZABLE[0]}'`, () => {
        registerQuestSystems(mockContainer);

        initializableQuestTokens.forEach(token => {
            expect(mockContainer.register).toHaveBeenCalledWith(
                token,
                expect.any(Function),
                expect.objectContaining({
                    tags: expect.arrayContaining(INITIALIZABLE),
                    lifecycle: 'singleton'
                })
            );
        });
    });

    it(`should NOT tag ${nonInitializableQuestTokens.length} services with '${INITIALIZABLE[0]}'`, () => {
        registerQuestSystems(mockContainer);

        nonInitializableQuestTokens.forEach(token => {
            // Need to find the specific call for this token
            const registrationCall = mockContainer.register.mock.calls.find(call => call[0] === token);
            expect(registrationCall).toBeDefined();
            // Check that tags array exists BUT does NOT contain INITIALIZABLE
            // Or, more simply, check if the options object *lacks* the tags array or the tag isn't in it
            const options = registrationCall[2];
            expect(options?.tags).not.toEqual(expect.arrayContaining(INITIALIZABLE));
            expect(options.lifecycle).toEqual('singleton'); // Still a singleton
        });
    });

    it('should log registration messages', () => {
        registerQuestSystems(mockContainer);

        expect(mockLogger.debug).toHaveBeenCalledWith('Quest Registration: Starting...');
        questTokens.forEach(token => {
            if (initializableQuestTokens.includes(token)) {
                expect(mockLogger.debug).toHaveBeenCalledWith(expect.stringContaining(`Registered ${String(token)} tagged with ${INITIALIZABLE[0]}`));
            } else {
                expect(mockLogger.debug).toHaveBeenCalledWith(expect.stringContaining(`Registered ${String(token)}.`));
            }
        });
        expect(mockLogger.info).toHaveBeenCalledWith(expect.stringContaining(`Quest Registration: Completed registering ${questTokens.length} quest services/systems.`));
    });


    it('should resolve QuestPrerequisiteService when QuestSystem is resolved', () => {
        // Arrange: Register necessary mocks for QuestSystem dependencies
        // Ensure ALL dependencies listed in QuestSystem's registration are mocked *and registered*
        mockContainer.register(tokens.EventBus, mockEventBus, {lifecycle: 'singleton'});
        mockContainer.register(tokens.EntityManager, mockEntityManager, {lifecycle: 'singleton'});
        mockContainer.register(tokens.GameStateManager, mockGameStateManager, {lifecycle: 'singleton'});
        mockContainer.register(tokens.QuestPrerequisiteService, mockQuestPrerequisiteService, {lifecycle: 'singleton'});
        mockContainer.register(tokens.QuestRewardService, mockQuestRewardService, {lifecycle: 'singleton'});
        mockContainer.register(tokens.ObjectiveEventListenerService, mockObjectiveEventListenerService, {lifecycle: 'singleton'});
        mockContainer.register(tokens.ObjectiveStateCheckerService, mockObjectiveStateCheckerService, {lifecycle: 'singleton'});
        // Register the *corrected* GameDataRepository mock
        mockContainer.register(tokens.GameDataRepository, mockGameDataRepository, {lifecycle: 'singleton'});
        // validatedEventDispatcher is needed by QuestRewardService factory, which is a dependency of QuestSystem factory
        mockContainer.register(tokens.ValidatedEventDispatcher, mockvalidatedEventDispatcher, {lifecycle: 'singleton'});

        // Call the registration function AFTER mocks are set up
        registerQuestSystems(mockContainer);

        // Spy on the container's resolve method *after* registration but *before* resolving QuestSystem
        const resolveSpy = jest.spyOn(mockContainer, 'resolve');

        // Act: Resolve QuestSystem, which should trigger its factory
        // This line should no longer throw the error about GameDataRepository
        const questSystemInstance = mockContainer.resolve(tokens.QuestSystem);

        // Assert: Check that the resolve spy was called with the dependency token
        // This assertion remains the primary goal of this specific test
        expect(resolveSpy).toHaveBeenCalledWith(tokens.QuestPrerequisiteService);

        // Also check other dependencies were resolved implicitly if needed for confidence
        expect(resolveSpy).toHaveBeenCalledWith(tokens.GameDataRepository); // Should have been called by the factory now
        expect(resolveSpy).toHaveBeenCalledWith(tokens.QuestRewardService);
        expect(resolveSpy).toHaveBeenCalledWith(tokens.EventBus);
        expect(resolveSpy).toHaveBeenCalledWith(tokens.EntityManager);
        expect(resolveSpy).toHaveBeenCalledWith(tokens.GameStateManager);
        expect(resolveSpy).toHaveBeenCalledWith(tokens.ObjectiveEventListenerService);
        expect(resolveSpy).toHaveBeenCalledWith(tokens.ObjectiveStateCheckerService);
        // Note: validatedEventDispatcher is resolved *indirectly* when QuestRewardService is resolved by the factory.
        // Spying might not catch it unless the mock resolve explicitly tracks nested calls.
        // But checking the direct dependencies of QuestSystem is usually sufficient here.

        expect(questSystemInstance).toBeDefined(); // Ensure resolution didn't fail overall

        // Cleanup spy
        resolveSpy.mockRestore();
    });

    it('should match snapshot for registration calls', () => {
        registerQuestSystems(mockContainer);
        const callsToSnapshot = mockContainer.register.mock.calls.filter(
            (call) => call[0] !== tokens.ILogger // Exclude pre-registered logger
        );
        // Remember to update snapshot if this is the first run after fixing mocks or logic: jest --updateSnapshot
        expect(callsToSnapshot).toMatchSnapshot();
    });

});