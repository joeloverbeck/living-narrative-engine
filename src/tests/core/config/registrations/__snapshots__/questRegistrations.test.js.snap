// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`registerQuestSystems should match snapshot for registration calls 1`] = `
[
  [
    "QuestPrerequisiteService",
    [Function],
    {
      "lifecycle": "singleton",
    },
  ],
  [
    "QuestRewardService",
    [Function],
    {
      "lifecycle": "singleton",
    },
  ],
  [
    "ObjectiveEventListenerService",
    [Function],
    {
      "lifecycle": "singleton",
    },
  ],
  [
    "ObjectiveStateCheckerService",
    [Function],
    {
      "lifecycle": "singleton",
    },
  ],
  [
    "QuestSystem",
    [Function],
    {
      "lifecycle": "singleton",
      "tags": [
        "initializableSystem",
      ],
    },
  ],
  [
    "QuestStartTriggerSystem",
    [Function],
    {
      "lifecycle": "singleton",
      "tags": [
        "initializableSystem",
      ],
    },
  ],
]
`;
